c10/cuda/CUDAStream.cpp

```diff
@@ -321,6 +334,43 @@ cudaStream_t CUDAStream::stream() const {
   return ptr->stream;
 }
 
+CUDAAssert* CUDAStream::assert_state() const {
+  auto ptr = CUDAStream_internals(*this);
+  AT_ASSERT(ptr);
+
+  // lazy create assert state
+  CUDAAssert* assert_state = ptr->assert_state.load();
+  if (assert_state == nullptr) {
+    // switch to device to associate host memory with it
+    CUDAGuard device_guard(device_index());
+
+    // allocate assert memory for this stream object
+    CUDAAssert* new_instance = nullptr;
+    C10_CUDA_CHECK(cudaHostAlloc(
+        (void**)&new_instance, sizeof(CUDAAssert), cudaHostAllocMapped));
+    memset(new_instance, 0, sizeof(CUDAAssert)); // clear memory
+    new_instance->mutex = new std::mutex();
```

Comment:
Technically this is UB, and you need to placement-new the mutex here.

Reply:
here mutex is just a pointer, if it was a regular member I would need placement new. I thought using placement for the whole CUDAAssert struct and use a regular mutex member but I decided against it because it is pinned memory, e.g. I thought it might make sense to use a simple plain C object.